app-ng/src/job.lisp

```diff
@@ -0,0 +1,99 @@
+(in-package #:qvm-app)
+
+(deftype job-status ()
+  '(member fraiche running finished interrupted error))
+
+(defstruct (job (:constructor %make-job))
+  "A job is an data-structure that permits a given function to be run in a threaded (and thread-safe) environment."
+  lock
+  threader
+  thread
+  %result                              ; I define this accessor below.
+  (status 'fraiche :type job-status)
+  error)
+
+(defmacro with-job-lock ((job &key (wait t)) &body body)
+  "Evaluate BODY with the LOCK acquired for the duration of BODY. If WAIT is T, block until the lock is available."
+  `(progn
+     (bt:acquire-lock (job-lock ,job) ,wait)
+     ,@body
+     (bt:release-lock (job-lock ,job))))
+
+(defmethod print-object ((job job) stream)
+  (print-unreadable-object (job stream :type nil :identity t)
+    (format stream "QVM Job")))
+
+(defun job-running-p (job)
+  "Is the JOB in the RUNNING state."
+  (with-job-lock (job)
+    (eq (job-status job) 'running)))
+
+(defun job-interrupted-p (job)
+  "Is the JOB in the INTERRUPTED state?"
+  (with-job-lock (job)
+    (eq (job-status job) 'interrupted)))
+
+(defun job-finished-p (job)
+  "Is the JOB in the FINISHED state?"
+  (with-job-lock (job)
+    (eq (job-status job) 'finished)))
+
+(defun make-job (fn)
+  "Make a JOB object which will, under JOB-START, run FN in a thread.
+
+The returned JOB can be repeatedly started if the JOB-STATUS is FINISHED (or INTERRUPTED). Only interact with the job after acquiring the lock (or alternatively by using WITH-JOB-LOCK)."
+  (let* ((job (%make-job))
+         (thr-fn (lambda ()
+                   (setf (job-thread job)
+                         (bt:make-thread
+                          (lambda ()
+                            (handler-case
+                                (progn
+                                  ;; Reset run-specific slots
+                                  (setf (job-%result job) nil)
+                                  (setf (job-error   job) nil)
+                                  (setf (job-status  job) 'running)
+                                  (setf (job-%result job) (funcall fn))
+                                  (setf (job-status  job) 'finished))
+                              (error (c)
+                                ;; TODO What happens to a thread when
+                                ;; there is an error in it?
+                                (setf (job-error job) c)
+                                (setf (job-status job) 'error)))))))))
+    (setf (job-threader job) thr-fn)
+    (setf (job-lock job) (bt:make-lock))
+    job))
+
+;;; CUSTODIAL
+
+(defun job-start (job)
```

Comment:
Suggestion: implement something like

```lisp
(defmacro define-job-action (action (job) &body body)
  `(defun ,action (,job)
     (with-job-lock (,job)
       ,@body)))
```
Then write `job-start`, etc. in terms of the above macro.